src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -791,6 +764,169 @@ public IEnumerable<TResult> OfType<TResult>()
+
+        /// <summary>
+        /// Inserts the specified values at the specified index.
+        /// </summary>
+        /// <param name="index">The index at which to insert the value.</param>
+        /// <param name="items">The elements to insert.</param>
+        /// <returns>The new immutable collection.</returns>
+        public ImmutableArray<T> InsertRange(int index, T[] items)
+        {
+            var self = this;
+            self.ThrowNullRefIfNotInitialized();
+            Requires.Range(index >= 0 && index <= self.Length, nameof(index));
+            Requires.NotNull(items, nameof(items));
+
+            if (items.Length == 0)
+            {
+                return self;
+            }
+            if (self.IsEmpty)
+            {
+                return new ImmutableArray<T>(items);
+            }
+
+            return self.InsertSpanRangeInternal(index, items);
+        }
+
+        /// <summary>
+        /// Inserts the specified values at the specified index.
+        /// </summary>
+        /// <param name="index">The index at which to insert the value.</param>
+        /// <param name="items">The elements to insert.</param>
+        /// <returns>The new immutable collection.</returns>
+        public ImmutableArray<T> InsertRange(int index, ReadOnlySpan<T> items)
+        {
+            var self = this;
+            self.ThrowNullRefIfNotInitialized();
+            Requires.Range(index >= 0 && index <= self.Length, nameof(index));
+
+            if (items.IsEmpty)
+            {
+                return self;
+            }
+            if (self.IsEmpty)
+            {
+                return items.ToImmutableArray();
+            }
+
+            return self.InsertSpanRangeInternal(index, items);
+        }
+
+        /// <summary>
+        /// Removes the specified values from this list.
+        /// </summary>
+        /// <param name="items">The items to remove if matches are found in this list.</param>
+        /// <param name="equalityComparer">
+        /// The equality comparer to use in the search.
+        /// </param>
+        /// <returns>
+        /// A new list with the elements removed.
+        /// </returns>
+        public ImmutableArray<T> RemoveRange(ReadOnlySpan<T> items, IEqualityComparer<T>? equalityComparer = null)
+        {
+            var self = this;
+            self.ThrowNullRefIfNotInitialized();
+
+            if (items.IsEmpty || self.IsEmpty)
+            {
+                return self;
+            }
+
+            if (items.Length == 1)
+            {
+                return self.Remove(items[0], equalityComparer);
+            }
+
+            var indicesToRemove = new SortedSet<int>();
+            foreach (var item in items)
+            {
+                int index = -1;
+                do
+                {
+                    index = self.IndexOf(item, index + 1, equalityComparer);
+                } while (index >= 0 && !indicesToRemove.Add(index) && index < self.Length - 1);
```

Comment:
This is O(M*N).  That's ok?

Reply:
@stephentoub Thanks! I would like to make it to O(M+N) but because existing `RemoveRange(some collection)` were also same O(M*N) logic which make me think here they would avoid allocating additional memory.

Anyway, I will make new and existing `RemoveRange `overloads be O(M+N) but hope can allocate some more memory :)  // cc @eiriktsarpalis

Reply:
Assuming you change it to _O(M+N)_ you might also be able to replace `SortedSet<int>` with a `List<int>` since indices would get added monotonically.

Reply:
Changed `RemoveRange(ReadOnlySpan)` (and other foreseen length of input items overloads) to be O(M+N), please review.

For existing `RemoveRange(IEnumerable)` method, it has to create hashing structure for `immutableArray` side rather than input items (compared with `ReadOnlySpan `overload) ? Then indices will not be added monotonically and we will need some sorted collection for indices?
If that, just confirm: should we change `RemoveRange(IEnumerable)` method also ?

Reply:
> should we change RemoveRange(IEnumerable) method also ?

Yes, I think we should, although I would make sure we add a few more tests to make sure no regressions are introduced (e.g. handling of null elements, duplicates, etc.)

Reply:
> > should we change RemoveRange(IEnumerable) method also ?
> 
> Yes, I think we should

Just suggest to separate it to another PR because it may not be related to current PR. I would like to implement it in new PR :)